File: overlay/python/hydra-kernel/src/main/python/hydra/dsl/meta/lib/regex.py
"""Phantom-typed term DSL for the hydra.lib.regex library."""

from hydra.typed import TypedTerm
from hydra.dsl.meta.phantoms import lib_primitive2 as primitive2, lib_primitive3 as primitive3


def find(pattern: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Find the first substring matching a regex pattern."""
    return primitive2(pattern, s)


def find_all(pattern: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Find all non-overlapping substrings matching a regex pattern."""
    return primitive2(pattern, s)


def matches(pattern: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Check whether an entire string matches a regex pattern."""
    return primitive2(pattern, s)


def replace(pattern: TypedTerm, replacement: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Replace the first occurrence of a regex pattern with a replacement string."""
    return primitive3(pattern, replacement, s)


def replace_all(pattern: TypedTerm, replacement: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Replace all non-overlapping occurrences of a regex pattern with a replacement string."""
    return primitive3(pattern, replacement, s)


def split(pattern: TypedTerm, s: TypedTerm) -> TypedTerm:
    """Split a string by a regex pattern."""
    return primitive2(pattern, s)

File: overlay/haskell/hydra-kernel/src/main/haskell/Hydra/Dsl/Meta/Lib/Regex.hs
-- | Phantom-typed term DSL for the hydra.lib.regex library

module Hydra.Dsl.Meta.Lib.Regex where

import Hydra.Typed
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import qualified Hydra.Dsl.Prims as Prims
import qualified Hydra.Lib.Regex as DefRegex


-- | Find the first substring matching a regex pattern.
find :: TypedTerm String -> TypedTerm String -> TypedTerm (Maybe String)
find = primitive2 (Prims.primName DefRegex.find)

-- | Find all non-overlapping substrings matching a regex pattern.
findAll :: TypedTerm String -> TypedTerm String -> TypedTerm [String]
findAll = primitive2 (Prims.primName DefRegex.findAll)

-- | Check whether an entire string matches a regex pattern.
matches :: TypedTerm String -> TypedTerm String -> TypedTerm Bool
matches = primitive2 (Prims.primName DefRegex.matches)

-- | Replace the first occurrence of a regex pattern with a replacement string.
replace :: TypedTerm String -> TypedTerm String -> TypedTerm String -> TypedTerm String
replace = primitive3 (Prims.primName DefRegex.replace)

-- | Replace all non-overlapping occurrences of a regex pattern with a replacement string.
replaceAll :: TypedTerm String -> TypedTerm String -> TypedTerm String -> TypedTerm String
replaceAll = primitive3 (Prims.primName DefRegex.replaceAll)

-- | Split a string by a regex pattern.
split :: TypedTerm String -> TypedTerm String -> TypedTerm [String]
split = primitive2 (Prims.primName DefRegex.split)
